bin/auth0setup/create.js

```diff
@@ -0,0 +1,49 @@
+var envLocation = '.env.local'
+var dotenv = require('dotenv')
+dotenv.config({path: envLocation})
+
+var fs = require('fs')
+var request = require('request');
+var makeClient = require('./makeClient.js')
+
+
+// add an assert statement. double check that the env file exists. throw and error if it doesn't exist.
+
+
+// get a token
+var options = { method: 'POST',
+  url: 'https://' + process.env.DOMAIN + '/oauth/token',
+  headers: { 'content-type': 'application/json' },
+  body: { 
+    grant_type: 'client_credentials',
+    client_id: process.env.EXPLORER_CLIENT_ID,
+    client_secret: process.env.EXPLORER_CLIENT_SECRET,
+    audience: 'https://' + process.env.DOMAIN + '/api/v2/'
+  },
+  json: true 
+}
+
+
+request(options, function (error, response, body) {
+  
+  if(error) throw error; // TODO: add human readable error message for bad url/dns error.
+	
+	
+  // overwrite .env file with access token
+  var envFile = fs.readFileSync(envLocation, 'utf8')
+  var envJson = dotenv.parse(envFile)
+  envJson.ACCESS_TOKEN = body.access_token
+  var envString = ''
+	
+  for(var k in envJson) {
+    envString = envString + k + '=' + envJson[k] + '\n'
+  }
+  fs.writeFileSync(envLocation, envString)
```

Comment:
@mugwump303 would you feel comfortable refactoring this code to collect all the variables from the created Auth0 APIs first, and then only write to the envFile once at the very end? I think this could be done really nicely if we use [Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) (maybe using [request-promise-native](https://github.com/request/request-promise-native), or maybe even just using something like node's [promisify utility](http://2ality.com/2017/05/util-promisify.html)) and [async await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function). I believe it could be done with something along these lines:

Changing requests like the one made in `makeClient` to have an exported function that returns a promise for the value(s) we care about. 
```js
// makeClient.js
const requestPromise = require('request-promise-native');
// ...
module.exports = async function makeClient(token) {
  // ...
  const response = await requestPromise(options);
  return response.client_id;
}
```

Then the final version of this file could look something like:
```js
async function setUpAuth0andSaveEnvironmentVariables(/* prompted variables */) {
  try {
    const token = await getAuth0Token(domain, explorerClientId, explorerClientSecret);
    const clientId = await makeClient(token);
    const apiIdentifier = await makeAPI(token);
    saveVariablesToEnvFile({
      REACT_APP_AUTH0_DOMAIN: domain,
      REACT_APP_AUTH0_CLIENT_ID: clientId,
      REACT_APP_AUTH0_API_IDENTIFIER: apiIdentifier,
    });
  } catch (err) {
    console.error('There was an error with the automated setup of auth0 for this project:');
    console.error(err);
    process.exit(1); // not sure if this is necessary or not
  }
}
// prompt for initial Auth0 keys to work with
setUpAuth0andSaveEnvironmentVariables(/* prompted variables */);
```

Reply:
It might take me a little more time, so I'll try to fix a few of the other things while looking into Promises and Aysnc/Await.